src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -837,26 +991,153 @@ public new CSharpCompilation ReplaceSyntaxTree(SyntaxTree oldTree, SyntaxTree ne
 
-            RemoveSyntaxTreeFromDeclarationMapAndTable(oldTree, ref declMap, ref declTable, ref referenceDirectivesChanged);
-            AddSyntaxTreeToDeclarationMapAndTable(newTree, _options, this.IsSubmission, ref declMap, ref declTable, ref referenceDirectivesChanged);
+            var numTreesRemoved = RemoveAllSyntaxTreesFromDeclarationMapAndTable(
+                oldTree,
+                newTree,
+                ref loadDirectiveMap,
+                ref declMap,
+                ref declTable,
+                ref referenceDirectivesChanged,
+                ref loadDirectivesChanged);
 
             var ordinalMap = this.syntaxTreeOrdinalMap;
-
-            Debug.Assert(ordinalMap.ContainsKey(oldTree)); // Checked by RemoveSyntaxTreeFromDeclarationMapAndTable
             var oldOrdinal = ordinalMap[oldTree];
+            ImmutableArray<SyntaxTree> newTrees;
+            if (loadDirectivesChanged)
+            {
+                Debug.Assert(!loadDirectiveMap.ContainsKey(oldTree));
+                Debug.Assert(!loadDirectiveMap.ContainsKey(newTree));
+
+                // If we're inserting new #load'ed trees, we'll rebuild
+                // the whole syntaxTree array and the ordinalMap.
+                var treesBuilder = ArrayBuilder<SyntaxTree>.GetInstance();
+                var ordinalMapBuilder = PooledDictionary<SyntaxTree, int>.GetInstance();
+
+                for (var i = 0; i < (oldOrdinal - numTreesRemoved); i++)
+                {
+                    var tree = _syntaxTrees[i];
+                    treesBuilder.Add(tree);
+                    ordinalMapBuilder.Add(tree, i);
+                }
+                AppendAllSyntaxTrees(treesBuilder, newTree, ordinalMapBuilder, ref loadDirectiveMap, ref declMap, ref declTable, ref referenceDirectivesChanged);
+                for (var i = oldOrdinal + 1; i < _syntaxTrees.Length; i++)
+                {
+                    var tree = _syntaxTrees[i];
+                    treesBuilder.Add(tree);
+                    ordinalMapBuilder.Add(tree, treesBuilder.Count);
+                }
+
+                newTrees = treesBuilder.ToImmutableAndFree();
+                ordinalMap = ordinalMapBuilder.ToImmutableDictionary();
+                ordinalMapBuilder.Free();
+                Debug.Assert(newTrees.Length == ordinalMap.Count);
+            }
+            else
+            {
+                AddSyntaxTreeToDeclarationMapAndTable(newTree, _options, this.IsSubmission, ref declMap, ref declTable, ref referenceDirectivesChanged);
+
+                Debug.Assert(ordinalMap.ContainsKey(oldTree)); // Checked by RemoveSyntaxTreeFromDeclarationMapAndTable
+
+                newTrees = _syntaxTrees.SetItem(oldOrdinal, newTree);
+
+                // CONSIDER: should this be an operation on ImmutableDictionary?
+                ordinalMap = ordinalMap.Remove(oldTree);
+                ordinalMap = ordinalMap.SetItem(newTree, oldOrdinal);
```

Comment:
Unless there's a performance penalty or something, I'd prefer to leave it as ```SetItem```, because that's what the previous code did (and I have no idea if there was a subtle reason for that).